aiida/tools/importexport/common/archive.py

```diff
@@ -217,9 +352,103 @@ def _read_json_file(self, filename):
+        """
+        try:
+            with io.open(self.folder.get_abs_path(filename), 'r', encoding='utf8') as fhandle:
+                return json.load(fhandle)
+        except OSError as error:
+            raise CorruptArchive('Error reading {}: {}'.format(filename, error))
+
+    @ensure_within_context
+    @ensure_unpacked
+    def _read_json_file_content(self, json_filename, key):
+        """Return value in JSON file
+
+        :param json_filename: Valid archive JSON file name.
+        :type json_filename: str
+        :param key: Key matching the value to be returned.
+        :type key: str
+
+        :return: Value of key in JSON file
+        :rtype: dict, str, list
+
+        :raises ~aiida.tools.importexport.common.exceptions.CorruptArchive: if the key cannot be found in JSON file
+        """
+        if json_filename == FILENAME_DATA:
+            json_file = self.data
+            mandatory_keys = self.MANDATORY_DATA_KEYS
+        elif json_filename == FILENAME_METADATA:
+            json_file = self.meta_data
+            mandatory_keys = self.MANDATORY_METADATA_KEYS
+        else:
+            raise CorruptArchive('{} is not a valid archive JSON file'.format(json_file))
+
+        if key in json_file:
+            return json_file[key]
+        if key not in mandatory_keys:
+            return {}
+        # else
+        raise CorruptArchive('{} is missing the mandatory key "{}"'.format(json_filename, key))
+
+    @ensure_within_context
+    def read_data(self, key):
+        """Return value in data.json
+
+        :param key: Key matching the value to be returned
+        :type key: str
+        """
+        return self._read_json_file_content(FILENAME_DATA, key)
+
+    @ensure_within_context
+    def read_meta_data(self, key):
+        """Return value in metadata.json
+
+        :param key: Key matching the value to be returned
+        :type key: str
         """
-        with io.open(self.folder.get_abs_path(filename), 'r', encoding='utf8') as fhandle:
-            return json.load(fhandle)
+        return self._read_json_file_content(FILENAME_METADATA, key)
+
+    @ensure_within_context
+    @ensure_unpacked
+    def _write_json_file(self, filename, data):
+        """Write the contents of a dictionary to a JSON file in the unpacked archive.
+
+        :param filename: the filename relative to the sandbox folder
+        :type filename: str
+        :param data: the data to be written to a JSON file
+        :type data: dict
+
+        :raises ~aiida.tools.importexport.common.exceptions.CorruptArchive: if there was an error writing the JSON file
+        """
+        try:
+            with io.open(self.folder.get_abs_path(filename), 'wb') as fhandle:
+                json.dump(data, fhandle)
+
+            self._keep = True
+        except OSError as error:
+            raise CorruptArchive('Error writing {}: {}'.format(filename, error))
```

Comment:
The `CorruptArchive` exception seems a bit weird to throw here. It is not that the original archive is corrupt it is just that writing the altered one fails. Should be something like an operational error